src/Schema/AST/ASTBuilder.php

```diff
@@ -225,20 +237,37 @@ protected function applyArgManipulators(): void
             if ($typeDefinition instanceof ObjectTypeDefinitionNode || $typeDefinition instanceof InterfaceTypeDefinitionNode) {
                 foreach ($typeDefinition->fields as $fieldDefinition) {
                     foreach ($fieldDefinition->arguments as $argumentDefinition) {
-                        foreach (
-                            $this->directiveLocator->associatedOfType($argumentDefinition, ArgManipulator::class)
-                            as $argManipulator
+                        /** @var array<string, bool> $executedManipulators */
+                        $executedManipulators = [];
+                        while (
+                            $argManipulator = $this->directiveLocator->associatedOfType($argumentDefinition, ArgManipulator::class)
+                                ->reject(fn (ArgManipulator $argManipulator): bool => isset($executedManipulators[$this->getManipulatorId($argManipulator)]))
+                                ->first()
                         ) {
                             $argManipulator->manipulateArgDefinition(
                                 $this->documentAST,
                                 $argumentDefinition,
                                 $fieldDefinition,
                                 $typeDefinition,
                             );
+
+                            $executedManipulators[$this->getManipulatorId($argManipulator)] = true;
                         }
                     }
                 }
             }
         }
     }
+
+    /**
+     * Return unique manipulator identifier.
+     */
+    protected function getManipulatorId(FieldManipulator|ArgManipulator|TypeManipulator $manipulator): string
+    {
+        if ($manipulator instanceof BaseDirective) {
+            return spl_object_hash($manipulator->directiveNode);
+        }
+
+        return get_class($manipulator);
```

Comment:
The different logic here is actually a huge unexpected gotcha, it basically forbids non-`BaseDirective` manipulators from being used repeatedly. Although unlikely, this is a breaking change.

If we don't find another solution, we can hide the repeated application of manipulators behind a configuration flag to make it opt-in for now. However, I would like to see a plan on how we could resolve this gotcha in an upcoming major version - perhaps by adding a method `directiveNode(): DirectiveNode` to the `Directive` interface?

Reply:
> since this feature would only add extra functionality everything will still work as before

I am saying it would not, so it will not. A schema that has multiple uses of the same non-`BaseDirective` at the same location would behave differently.